src/initio/initio.c

```diff
@@ -0,0 +1,794 @@
+      mraa_gpio_close(dev);
+      return NULL;
+    }  else {
+      if (++idx == n) return dev;
+    }
+  }
+
+  /* Check for edge. */
+  int edge = -1;
+  if (strncmp(proto[idx], G_EDGE_NONE, strlen(G_EDGE_NONE)) == 0) {
+    edge = MRAA_GPIO_EDGE_NONE;
+  } else if (strncmp(proto[idx], G_EDGE_BOTH, strlen(G_EDGE_BOTH)) == 0) {
+    edge = MRAA_GPIO_EDGE_BOTH;
+  } else if (strncmp(proto[idx], G_EDGE_RISING, strlen(G_EDGE_RISING)) == 0) {
+    edge = MRAA_GPIO_EDGE_RISING;
+  } else if (strncmp(proto[idx], G_EDGE_FALLING, strlen(G_EDGE_FALLING)) == 0) {
+    edge = MRAA_GPIO_EDGE_FALLING;
+  } else {}
+
+  if (edge != -1) {
+    if (mraa_gpio_dir(dev, (mraa_gpio_edge_t)edge) != MRAA_SUCCESS) {
+      syslog(LOG_ERR, "parse_gpio: error setting up gpio edge %s", proto[idx]);
+      mraa_gpio_close(dev);
+      return NULL;
+    }  else {
+      if (++idx == n) return dev;
+    }
+  }
+
+  /* Check for input mode. */
+  int input_mode = -1;
+  if (strncmp(proto[idx], G_INPUT_ACTIVE_HIGH, strlen(G_INPUT_ACTIVE_HIGH)) == 0) {
+    input_mode = MRAA_GPIO_ACTIVE_HIGH;
+  } else if (strncmp(proto[idx], G_INPUT_ACTIVE_LOW, strlen(G_INPUT_ACTIVE_LOW)) == 0) {
+    input_mode = MRAA_GPIO_ACTIVE_LOW;
+  } else {}
+
+  if (input_mode != -1) {
+    if (mraa_gpio_input_mode(dev, (mraa_gpio_input_mode_t)input_mode) != MRAA_SUCCESS) {
+      syslog(LOG_ERR, "parse_gpio: error setting up gpio input mode %s", proto[idx]);
+      mraa_gpio_close(dev);
+      return NULL;
+    }  else {
+      if (++idx == n) return dev;
+    }
+  }
+
+  /* Check for driver output mode. */
+  int driver_mode = -1;
+  if (strncmp(proto[idx], G_OUTPUT_OPEN_DRAIN, strlen(G_OUTPUT_OPEN_DRAIN)) == 0) {
+    driver_mode = MRAA_GPIO_OPEN_DRAIN;
+  } else if (strncmp(proto[idx], G_OUTPUT_PUSH_PULL, strlen(G_OUTPUT_PUSH_PULL)) == 0) {
+    driver_mode = MRAA_GPIO_PUSH_PULL;
+  } else {}
+
+  if (driver_mode != -1) {
+    if (mraa_gpio_input_mode(dev, (mraa_gpio_out_driver_mode_t)driver_mode) != MRAA_SUCCESS) {
+      syslog(LOG_ERR, "parse_gpio: error setting up gpio driver mode %s", proto[idx]);
+      mraa_gpio_close(dev);
+      return NULL;
+    }
+  }
+
+  return dev;
+}
+
+mraa_result_t mraa_io_init(const char* strdesc, mraa_io_descriptor **desc)
+{
+  mraa_result_t status = MRAA_SUCCESS;
+  char *p_str = NULL;
+  /* Allocate space for the descriptor */
+  mraa_io_descriptor *new_desc = calloc(1, sizeof(mraa_io_descriptor));
+  if (new_desc == NULL) {
+    syslog(LOG_ERR, "mraa_io_init: Failed to allocate memory for context");
+    desc = NULL;
+    return MRAA_ERROR_NO_RESOURCES;
+  }
+
+  /* First, get rid of the leftover string. */
+  int mraa_upm_num_str = 0;
+  char **mraa_upm_str = mraa_tokenize_string(strdesc, MRAA_UPM_SEP, &mraa_upm_num_str);
+  if (mraa_upm_num_str == 0) {
+    syslog(LOG_ERR, "mraa_io_init: Failed to allocate memory for context");
+    free(new_desc);
+    return MRAA_ERROR_INVALID_PARAMETER;
```

Comment:
If our error message just before this is "failed to allocate memory", then `MRAA_ERROR_NO_RESOURCES` is more appropriate here. However the error message looks wrong to me - if we've got 0 out of `mraa_tokenize_string()`, there's simply no such delimiter there, right? And that has nothgin to do with allocations. Please double-check.